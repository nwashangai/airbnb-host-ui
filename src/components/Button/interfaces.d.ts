export interface ButtonProps {
  size?: string;
  type?: string;
  icon?: string;
  name: string;
  disabled?: boolean;
  onClick?: any;
  submit?: boolean;
  id?: string;
  isActive?: boolean;
  classes?: string;
  alignImageRight?: boolean;
  isLoading?: boolean;
}
