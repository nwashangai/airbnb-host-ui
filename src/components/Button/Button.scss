@import 'sass-rem';
@import 'assets/variables';

%generic-btn {
  background-color: map-get($grey, 100);
  width: rem(78px);
  height: rem(28px);
  border: rem(1px) solid map-get($palette, primary);
  border-radius: rem(4px);
  font-size: rem(14px);
  font-weight: 600;
  box-shadow: 0 rem(2px) rem(9px) 0 rgba(0, 0, 0, 0.03);
  cursor: pointer;
  color: map-get($palette, primary);
  letter-spacing: 0.03rem;
  outline: none;

  img {
    margin-left: rem(8px);
  }
}

.inline {
  position: relative;

  &::before {
    top: 50%;
  }
}

.button {
  &.small {
    @extend %generic-btn;
  }

  &.back {
    background-color: map-get($grey, 2000);
    width: 2rem;
    height: 2rem;
    border: hidden;
    outline: none;
  }

  &.disabled {
    width: 7.625rem;
    height: 2.5rem;
    font-size: 1rem;
    font-weight: 700;
    cursor: not-allowed;
    opacity: 0.8;
  }

  &.regular {
    @extend %generic-btn;
    width: 100%;
    height: rem(40px);
    max-height: rem(40px);
    padding: rem(5px) rem(20px);

    img {
      margin: 0;
    }

    &--bold {
      font-weight: 700;
    }

    &--light {
      font-weight: 300;
    }

    &.blue,
    &.active {
      background-color: map-get($palette, primary);
      color: map-get($grey, 100);
    }
  }

  &.yellow {
    @extend %generic-btn;
    background-color: map-get($grey, 100);
    width: rem(108px);
    height: 2.5rem;
    border: rem(1px) solid map-get($palette, secondary);
    border-radius: rem(2px);
    font-weight: 500;
    color: map-get($palette, secondary);

    &.active {
      background-color: map-get($palette, secondary);
      color: map-get($grey, 100);
    }
  }

  &.load-more-button {
    @extend %generic-btn;
    width: rem(118px);
    height: rem(42);
    box-sizing: border-box;
    border: rem(1px) solid map-get($grey, 900);
    border-radius: rem(2px);
  }

  &.analytics-filter-button {
    display: flex;
    justify-content: space-between;
    background-color: map-get($grey, 100);
    width: rem(161px);
    height: rem(41px);
    padding: 0 rem(15px);
    box-sizing: border-box;
    border: rem(1px) solid map-get($grey, 900);
    border-radius: rem(4px);
    font-size: rem(16px);
    font-weight: 300;
    box-shadow: none;
    color: map-get($palette, settings-menu);
  }
}
