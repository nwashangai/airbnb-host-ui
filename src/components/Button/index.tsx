// react libraries
import * as React from 'react';

// styles
import './Button.scss';

// component
import Spinner from 'components/Spinner';

// interfaces
import { ButtonProps } from './interfaces';

const Button = (props: ButtonProps) => {
  let {
    size,
    type,
    isActive,
    submit = false,
    isLoading = false,
    alignImageRight = true,
    disabled,
    classes,
    ...rest
  } = props;
  type = !type ? 'hollow' : type;
  size = !size ? 'regular' : size;
  isActive = isActive !== undefined ? isActive : false;

  const classNames = isActive
    ? `button active ${size} ${type}`
    : `button ${size} ${type}`;

  const renderButtonContent = (props: any) => {
    // spinner shows when isLoading is true
    if (isLoading) {
      return <Spinner />;
    }

    return alignImageRight ? (
      <React.Fragment>
        {props.name} {props.icon && <img src={props.icon} alt="" />}
      </React.Fragment>
    ) : (
      <React.Fragment>
        {props.icon && <img src={props.icon} alt="" />} {props.name}
      </React.Fragment>
    );
  };

  const classList = `${
    classes ? `${classNames} ${classes}` : `${classNames}`
  } ${disabled ? 'disabled' : ''}`.trim();

  return (
    <button
      {...rest}
      type={submit ? 'submit' : 'button'}
      id={props.id}
      disabled={disabled}
      className={classList}
      onClick={props.onClick}
    >
      {renderButtonContent(props)}
    </button>
  );
};

export default Button;
