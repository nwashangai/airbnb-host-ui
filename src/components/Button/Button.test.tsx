// react libraries
import * as React from 'react';

// third-party libraries
import { mount } from 'enzyme';

// components
import Button from './';

describe('The button component', () => {
  const TEST_ICON = 'test-icon';
  const TEST_NAME = 'test-name';
  const TEST_TYPE = 'test-type';
  const TEST_SIZE = 'test-size';
  const TEST_ISACTIVE = true;
  const TEST_CLASSES = 'preview-form';

  it('displays a button with classes "button regular hollow" when mounted with no configuration props', () => {
    const wrapper = mount(<Button name={TEST_NAME} />);
    const button = wrapper.find('button');

    expect(button).toMatchSnapshot();
    expect(button.props().className).toBe('button regular hollow');
  });

  it('displays an image if the icon prop is passed', () => {
    const wrapper = mount(<Button name="test-name" icon={TEST_ICON} />);
    const img = wrapper.find('img');

    expect(wrapper).toMatchSnapshot();
    expect(img.props().src).toBe(TEST_ICON);
  });

  it('overwrites type and size props when values are provided', () => {
    const wrapper = mount(
      <Button
        name={TEST_NAME}
        type={TEST_TYPE}
        size={TEST_SIZE}
        icon={TEST_ICON}
      />
    );
    const button = wrapper.find('button');

    expect(wrapper).toMatchSnapshot();
    expect(button.props().className).toBe(`button ${TEST_SIZE} ${TEST_TYPE}`);
  });

  it('adds an active class to the button when isActive props is passed', () => {
    const wrapper = mount(<Button name={TEST_NAME} isActive={TEST_ISACTIVE} />);
    const button = wrapper.find('button');

    expect(wrapper).toMatchSnapshot();
    expect(button.props().className).toBe(`button active regular hollow`);
  });

  it('spreads additional props passed into the component down into the button element', () => {
    const wrapper = mount(
      <Button disabled name={TEST_NAME} isActive={TEST_ISACTIVE} />
    );
    const button = wrapper.find('button');

    expect(button.props().disabled).toBe(true);
  });

  it('adds the additional classes passed into the button', () => {
    const wrapper = mount(<Button name={TEST_NAME} classes={TEST_CLASSES} />);
    expect(wrapper.find('button').props().className).toMatch(TEST_CLASSES);
  });

  it('displays the image tag to the right if the alignImageRight prop is passed as false', () => {
    const wrapper = mount(
      <Button
        name={TEST_NAME}
        icon="images/fake.svg"
        alignImageRight={false}
        classes={TEST_CLASSES}
      />
    );
    expect(wrapper.find('button').html()).toMatch(
      `<img src="images/fake.svg" alt=""> test-name`
    );
  });

  it('displays the image tag to the left if the alignImageRight prop is not passed', () => {
    const wrapper = mount(<Button name={TEST_NAME} classes={TEST_CLASSES} />);
    expect(wrapper.find('button').html()).toMatch(`test-name`);
  });

  it('should not display spinner when isLoading is false', () => {
    const wrapper = mount(<Button name={TEST_NAME} icon={TEST_ICON} />);
    expect(wrapper.find('Spinner')).toHaveLength(0);
  });

  it('displays spinner when isLoading is true', () => {
    const wrapper = mount(
      <Button name={TEST_NAME} icon={TEST_ICON} isLoading={true} />
    );
    expect(wrapper.find('.spinner')).toHaveLength(1);
  });

  it('displays the image tag to the left if the alignImageRight prop is passed as true', () => {
    const wrapper = mount(
      <Button
        name="export"
        icon="icon.svg"
        alignImageRight={true}
        classes={TEST_CLASSES}
      />
    );

    expect(wrapper.find('button').html()).toMatch(
      `<button name="export" icon="icon.svg" type="button" class="button regular hollow preview-form">export <img src="icon.svg" alt=""></button>`
    );
  });
});
