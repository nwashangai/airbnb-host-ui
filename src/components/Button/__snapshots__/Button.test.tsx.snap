// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The button component adds an active class to the button when isActive props is passed 1`] = `ReactWrapper {}`;

exports[`The button component displays a button with classes "button regular hollow" when mounted with no configuration props 1`] = `ReactWrapper {}`;

exports[`The button component displays an image if the icon prop is passed 1`] = `ReactWrapper {}`;

exports[`The button component overwrites type and size props when values are provided 1`] = `ReactWrapper {}`;
